src/modsim2/similarity/metrics/kde.py

```diff
@@ -0,0 +1,303 @@
+import logging
+
+import numpy as np
+from scipy import integrate
+from sklearn.neighbors import KernelDensity, NearestNeighbors
+
+from .metrics import DistanceMetric
+
+# Set module logger
+logger = logging.getLogger(__name__)
+
+
+class KDE(DistanceMetric):
+    """
+    Class to calculate the distance between two datasets
+    using kernel density estimation (KDE). The density estimation
+    is calculated for the source and target (A and B) datasets.
+    A metric is then applied to calculate a distance between the
+    density distributions.
+
+    sklearn's KernelDensity class has been used as a variety of
+    kernels are implemented and the bandwidth can be estimated
+    using Scott's Rule.
+    """
+
+    def __init__(self, seed: int) -> None:
+        super().__init__(seed)
+
+    def _pre_process_data(
+        self,
+        data_A: np.ndarray,
+        data_B: np.ndarray,
+        embedding_name: str,
+        embedding_kwargs: dict,
+    ) -> tuple[np.ndarray, np.ndarray]:
+
+        # Embed the data
+        embed_A, embed_B = self._embed_data(
+            data_A=data_A,
+            data_B=data_B,
+            embedding_name=embedding_name,
+            embedding_kwargs=embedding_kwargs,
+        )
+
+        return embed_A, embed_B
+
+    @staticmethod
+    def _kde_distance(
+        num_dimensions: int,
+        func: callable,
+        integration_kwargs: dict,
+    ) -> tuple[float, float]:
+        """
+        Integrates a given function over a number of dimensions. Returns the result
+        of the integration
+
+        Args:
+            num_dimensions : the number of dimensions that are to be integrated over
+            func: the function to be integrated
+            integration_kwargs: A dictionary of key word arguments to be passed to the
+                            integration function
+
+        Returns:
+            distance: the result of the integration
+            abs_error: the maximum of the estimates of the absolute error in the
+                    various integration results.
+        """
+
+        # the bounds over which the integration will be performed
+        bounds = [[-np.inf, np.inf] for _ in range(num_dimensions)]
+
+        # perform integration, returns distance and absolute error
+        distance, abs_error = integrate.nquad(func, bounds, **integration_kwargs)
+
```

Comment:
We might want to analyse or report the error being returned from the integration. It can also help if we need to tune the parameters (particularly if we want to reduce the 'limit' param to help the code run faster)

Reply:
Am I right in thinking this currently logs a warning for the problems we've experienced?
How would you suggest we do the analysis/reporting of errors?